### ISSUE TITLE
kapp should be able to adopt resource with vesioned annotation

### ISSUE THREAD

**Issue by cari-lynn** [k14s/kapp#156]
kapp should be able to adopt resources annotated with `kapp.k14s.io/versioned: ""` that are deployed previously with a tool other than kapp. Currently kapp panics.
kapp could 'adopt' a resource by deleting the one currently in the cluster and creating a new one.

Reproduction steps:
1. Deploy a [versioned resource](https://github.com/k14s/kapp/blob/7cc75a5363584520d2bc8a7bd021c8350716edd3/docs/diff.md#versioned-resources) with `kubectl`
2. Deploy the versioned resource again with `kapp`
3. kapp results in this error:
```
panic: Missing template version

goroutine 1 [running]:
github.com/k14s/kapp/pkg/kapp/diff.TemplateResource.Version(0x20c3440, 0xc0004121e0, 0x0, 0x0, 0x0, 0x1)
	github.com/k14s/kapp@/pkg/kapp/diff/template_resource.go:40 +0x153
github.com/k14s/kapp/pkg/kapp/diff.ChangeSetWithTemplates.assignNewNames(0xc00007b9a0, 0x1, 0x1, 0xc000539620, 0x1, 0x1, 0xc00010c7e0, 0x2, 0x2, 0x1, ...)
	github.com/k14s/kapp@/pkg/kapp/diff/change_set_with_templates.go:105 +0x196
github.com/k14s/kapp/pkg/kapp/diff.ChangeSetWithTemplates.Calculate(0xc00007b9a0, 0x1, 0x1, 0xc000539620, 0x1, 0x1, 0xc00010c7e0, 0x2, 0x2, 0x1, ...)
	github.com/k14s/kapp@/pkg/kapp/diff/change_set_with_templates.go:37 +0x19c
github.com/k14s/kapp/pkg/kapp/cmd/app.(*DeployOptions).calculateAndPresentChanges(0xc0002b7200, 0xc00007b9a0, 0x1, 0x1, 0xc000539620, 0x1, 0x1, 0xc0004229c0, 0x1, 0x1, ...)
	github.com/k14s/kapp@/pkg/kapp/cmd/app/deploy.go:311 +0x3d6
github.com/k14s/kapp/pkg/kapp/cmd/app.(*DeployOptions).Run(0xc0002b7200, 0x0, 0x0)
	github.com/k14s/kapp@/pkg/kapp/cmd/app/deploy.go:136 +0x7db
github.com/k14s/kapp/pkg/kapp/cmd/app.NewDeployCmd.func1(0xc0002eb680, 0xc0002fc9c0, 0x0, 0x4, 0xc0003c5c80, 0xc000127be0)
	github.com/k14s/kapp@/pkg/kapp/cmd/app/deploy.go:48 +0x2a
github.com/k14s/kapp/pkg/kapp/cmd.NewKappCmd.func1.1(0xc0002eb680, 0xc0002fc9c0, 0x0, 0x4, 0x0, 0x0)
	github.com/k14s/kapp@/pkg/kapp/cmd/kapp.go:124 +0xcc
github.com/cppforlife/cobrautil.ReconfigureLeafCmd.func1(0xc0002eb680, 0xc0002fc9c0, 0x0, 0x4, 0x0, 0x0)
	github.com/cppforlife/cobrautil@v0.0.0-20200108171912-b289811fd2b9/misc.go:67 +0x135
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc0002eb680, 0xc0002fc9c0, 0x0, 0x4, 0x0, 0x0)
	github.com/cppforlife/cobrautil@v0.0.0-20200108171912-b289811fd2b9/misc.go:25 +0xaf
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc0002eb680, 0xc0002fc9c0, 0x0, 0x4, 0x0, 0x0)
	github.com/cppforlife/cobrautil@v0.0.0-20200108171912-b289811fd2b9/misc.go:25 +0xaf
github.com/spf13/cobra.(*Command).execute(0xc0002eb680, 0xc0002fc940, 0x4, 0x4, 0xc0002eb680, 0xc0002fc940)
	github.com/spf13/cobra@v0.0.3/command.go:762 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002eac80, 0xc0002eac80, 0xc0002fd7c0, 0x103b33a)
	github.com/spf13/cobra@v0.0.3/command.go:852 +0x2ea
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v0.0.3/command.go:800
main.main()
	github.com/k14s/kapp@/cmd/kapp/kapp.go:27 +0x16f
```

**Comment by Zebradil** [comment]
In my case kapp did panic on resource which was not annotated with `kapp.k14s.io/versioned: ""`:

```
panic: Missing template version in resource 'secret/oauth2-proxy (v1) namespace: oauth2-proxy'

goroutine 1 [running]:
github.com/k14s/kapp/pkg/kapp/diff.TemplateResource.Version(0x15be1c0, 0xc000010000, 0x0, 0x0, 0x0, 0x1)
        github.com/k14s/kapp@/pkg/kapp/diff/template_resource.go:43 +0x1e8
github.com/k14s/kapp/pkg/kapp/diff.ChangeSetWithTemplates.assignNewNames(0xc000646250, 0x1, 0x1, 0xc0003a05b0, 0x1, 0x1, 0xc00009a1e0, 0x2, 0x2, 0x1, ...)
        github.com/k14s/kapp@/pkg/kapp/diff/change_set_with_templates.go:108 +0x196
github.com/k14s/kapp/pkg/kapp/diff.ChangeSetWithTemplates.Calculate(0xc000646250, 0x1, 0x1, 0xc0003a05b0, 0x1, 0x1, 0xc00009a1e0, 0x2, 0x2, 0x1, ...)
        github.com/k14s/kapp@/pkg/kapp/diff/change_set_with_templates.go:40 +0x19c
github.com/k14s/kapp/pkg/kapp/cmd/app.(*DeployOptions).calculateAndPresentChanges(0xc00012cf00, 0xc000646250, 0x1, 0x1, 0xc0003a05b0, 0x1, 0x1, 0xc0006f8ea0, 0x1, 0x1, ...)
        github.com/k14s/kapp@/pkg/kapp/cmd/app/deploy.go:325 +0x3ee
github.com/k14s/kapp/pkg/kapp/cmd/app.(*DeployOptions).Run(0xc00012cf00, 0x0, 0x0)
        github.com/k14s/kapp@/pkg/kapp/cmd/app/deploy.go:143 +0x7cd
github.com/k14s/kapp/pkg/kapp/cmd/app.NewDeployCmd.func1(0xc00012d180, 0xc0000d8150, 0x0, 0x7, 0xc0003aa580, 0xc0004a7bc8)
        github.com/k14s/kapp@/pkg/kapp/cmd/app/deploy.go:54 +0x2a
github.com/k14s/kapp/pkg/kapp/cmd.NewKappCmd.func1.1(0xc00012d180, 0xc0000d8150, 0x0, 0x7, 0x0, 0x0)
        github.com/k14s/kapp@/pkg/kapp/cmd/kapp.go:127 +0xcc
github.com/cppforlife/cobrautil.ReconfigureLeafCmd.func1(0xc00012d180, 0xc0000d8150, 0x0, 0x7, 0x0, 0x0)
        github.com/cppforlife/cobrautil@v0.0.0-20200108171912-b289811fd2b9/misc.go:67 +0x135
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc00012d180, 0xc0000d8150, 0x0, 0x7, 0x0, 0x0)
        github.com/cppforlife/cobrautil@v0.0.0-20200108171912-b289811fd2b9/misc.go:25 +0xaf
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc00012d180, 0xc0000d8150, 0x0, 0x7, 0x0, 0x0)
        github.com/cppforlife/cobrautil@v0.0.0-20200108171912-b289811fd2b9/misc.go:25 +0xaf
github.com/spf13/cobra.(*Command).execute(0xc00012d180, 0xc0000d8070, 0x7, 0x7, 0xc00012d180, 0xc0000d8070)
        github.com/spf13/cobra@v0.0.3/command.go:762 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000e8c80, 0xc0000e8c80, 0xc000358f00, 0x10bc432)
        github.com/spf13/cobra@v0.0.3/command.go:852 +0x2ea
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v0.0.3/command.go:800
main.main()
        github.com/k14s/kapp@/cmd/kapp/kapp.go:31 +0x175
```

Could this case be handled in the same way as proposed in the first comment (when kapp just recreates the resource)?

**Comment by cppforlife** [comment]
@Zebradil im curious how you ended up with `secret/oauth2-proxy` having versioned annotation? did you also deploy it with kubectl before using kapp?

**Comment by Zebradil** [comment]
Yes, initially the secret was deployed with `kubectl` without versioned annotation. Afterwards I decided to switch to `kapp`: added the annotation and tried to deploy with `kapp`.

**Comment by renuy** [comment]
Available in release v0.39.0
